File: face/src/state/templates.ts
import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";

// The 7 session-lifecycle templates PromptBuilder.Render() looks for under <planDir>/*.md
// (Core/PromptBuilder.cs) — if one isn't on disk, the engine falls back to its own built-in
// default (not duplicated here; editing here always creates/overrides the file on disk).
export const SESSION_TEMPLATES = ["session.md", "fix.md", "resume.md", "advisor.md", "audit.md", "review.md", "verify.md"];

export interface TemplateEntry {
  /** Display name, e.g. "session.md" or "personas/architect.md". */
  label: string;
  /** Absolute path on disk. */
  path: string;
  exists: boolean;
}

/** Lists the known session templates + whatever's in <planDir>/personas — direct filesystem
 * access, no HTTP round-trip: the Face runs on the same machine as the engine (D1), and
 * PromptBuilder/PersonaRegistry already re-read these files from disk on every render, so editing
 * them here needs no new engine primitive (design doc D11). */
export function listTemplates(planDir: string): TemplateEntry[] {
  const entries: TemplateEntry[] = SESSION_TEMPLATES.map((name) => {
    const path = join(planDir, name);
    return { label: name, path, exists: existsSync(path) };
  });

  const personasDir = join(planDir, "personas");
  if (existsSync(personasDir)) {
    try {
      for (const f of readdirSync(personasDir)) {
        if (!f.endsWith(".md")) continue;
        const path = join(personasDir, f);
        entries.push({ label: `personas/${f}`, path, exists: true });
      }
    } catch {
      /* best effort — a permissions error here shouldn't crash the editor list */
    }
  }
  return entries;
}

export function readTemplate(path: string): string {
  if (!existsSync(path)) return "";
  try {
    return readFileSync(path, "utf8");
  } catch {
    return "";
  }
}

export function writeTemplate(path: string, content: string): void {
  const dir = dirname(path);
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
  writeFileSync(path, content, "utf8");
}

File: face-go/internal/templates/templates.go
// Package templates lists and edits the session-lifecycle prompt/persona templates that
// PromptBuilder/PersonaRegistry re-read from disk on every render (Core/PromptBuilder.cs). The
// Face runs on the same machine as the engine, so this is direct filesystem access — no HTTP
// round-trip and no new engine primitive needed. Saving here takes effect at the next session
// boundary, exactly like hand-editing the file would.
package templates

import (
	"os"
	"path/filepath"
	"strings"
)

// SessionTemplates are the 7 files PromptBuilder.Render() looks for under <planDir>/*.md. If one
// isn't on disk, the engine falls back to its own built-in default.
var SessionTemplates = []string{
	"session.md", "fix.md", "resume.md",
	"advisor.md", "audit.md", "review.md",
	"verify.md",
}

type Entry struct {
	// Label is the display name, e.g. "session.md" or "personas/architect.md".
	Label  string
	Path   string
	Exists bool
}

// List returns the known session templates plus whatever's under <planDir>/personas.
func List(planDir string) []Entry {
	entries := make([]Entry, 0, len(SessionTemplates))
	for _, name := range SessionTemplates {
		path := filepath.Join(planDir, name)
		_, err := os.Stat(path)
		entries = append(entries, Entry{Label: name, Path: path, Exists: err == nil})
	}

	personasDir := filepath.Join(planDir, "personas")
	if fis, err := os.ReadDir(personasDir); err == nil {
		for _, fi := range fis {
			if fi.IsDir() || !strings.HasSuffix(fi.Name(), ".md") {
				continue
			}
			entries = append(entries, Entry{
				Label:  "personas/" + fi.Name(),
				Path:   filepath.Join(personasDir, fi.Name()),
				Exists: true,
			})
		}
	}
	return entries
}

func Read(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return string(data)
}

func Write(path, content string) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, []byte(content), 0o644)
}
